### PR TITLE
improvement: introduce lock free queue for `global_task_queue`

### LINKED ISSUE #4817: Introduce lock-free queue for `global_tasks_queue` in processor
**Summary**

Introduce lock-free queue for `global_tasks_queue` in processor.

Current, `global_tasks_queue` in `PipelineExecutor` protectd by Mutex, this is likely to be a race contention.

cc @sundy-li @zhang2014

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog

- Performance Improvement

## Related Issues

Fixes #4817

### CODE REVIEW

File: query/src/pipelines/new/executor/executor_tasks.rs [hunk 1]
```diff
@@ -53,17 +54,14 @@ impl ExecutorTasksQueue {
     /// Method is thread unsafe and require thread safe call
     pub unsafe fn steal_task_to_context(&self, context: &mut ExecutorWorkerContext) {
         {
-            let mut workers_tasks = self.workers_tasks.lock();
-
-            if !workers_tasks.is_empty() {
-                let task = workers_tasks.pop_task(context.get_worker_num());
+            if !self.workers_tasks.is_empty() {
+                let task = self.workers_tasks.pop_task(context.get_worker_num());
```
Comment (zhang2014) [inline query/src/pipelines/new/executor/executor_tasks.rs:58]:
``` rust
if !self.workers_tasks.is_empty() {
                let task = self.workers_tasks.pop_task(context.get_worker_num());
```
Maybe this is thread unsafe?

File: query/src/pipelines/new/executor/executor_tasks.rs [hunk 2]
```diff
@@ -53,17 +54,14 @@ impl ExecutorTasksQueue {
...
                 context.set_task(task);
 
                 let workers_notify = context.get_workers_notify();
-                if !workers_tasks.is_empty() && !workers_notify.is_empty() {
+                if !self.workers_tasks.is_empty() && !workers_notify.is_empty() {
                     let worker_id = context.get_worker_num();
-                    let wakeup_worker_id = workers_tasks.best_worker_id(worker_id + 1);
-                    drop(workers_tasks);
+                    let wakeup_worker_id = self.workers_tasks.best_worker_id(worker_id + 1);
```
Comment (zhang2014) [inline query/src/pipelines/new/executor/executor_tasks.rs:64]:
``` rust
if !self.workers_tasks.is_empty() && !workers_notify.is_empty()

let wakeup_worker_id = self.workers_tasks.best_worker_id(worker_id + 1);
```

thread unsafe.

File: query/src/pipelines/new/executor/executor_tasks.rs [hunk 3]
```diff
@@ -116,11 +112,9 @@ impl ExecutorTasksQueue {
     }
 
     pub fn completed_async_task(&self, task: CompletedAsyncTask) {
-        let mut workers_tasks = self.workers_tasks.lock();
-
         let worker_id = task.worker_id;
-        workers_tasks.tasks_size += 1;
-        workers_tasks.workers_completed_async_tasks[worker_id].push_back(task);
+        self.workers_tasks.tasks_size.fetch_add(1, Ordering::SeqCst);
+        self.workers_tasks.workers_completed_async_tasks[worker_id].push(task);
```
Comment (zhang2014) [inline query/src/pipelines/new/executor/executor_tasks.rs:117]: Thread unsafe.

File: query/src/pipelines/new/executor/executor_tasks.rs [hunk 4]
```diff
@@ -156,34 +150,34 @@ impl ExecutorTasks {
...
 
     pub fn is_empty(&self) -> bool {
-        self.tasks_size == 0
+        self.tasks_size.load(Ordering::Relaxed) == 0
     }
 
     #[inline]
-    fn pop_worker_task(&mut self, worker_id: usize) -> ExecutorTask {
-        if let Some(processor) = self.workers_sync_tasks[worker_id].pop_front() {
+    fn pop_worker_task(&self, worker_id: usize) -> ExecutorTask {
```
Comment (zhang2014) [inline query/src/pipelines/new/executor/executor_tasks.rs:171]: We can't guarantee that this method is thread safe.

File: query/src/pipelines/new/executor/executor_tasks.rs [hunk 5]
```diff
@@ -232,17 +226,17 @@ impl ExecutorTasks {
         ExecutorTask::None
     }
 
-    pub unsafe fn push_task(&mut self, worker_id: usize, task: ExecutorTask) {
-        self.tasks_size += 1;
-        let sync_queue = &mut self.workers_sync_tasks[worker_id];
-        let async_queue = &mut self.workers_async_tasks[worker_id];
-        let completed_queue = &mut self.workers_completed_async_tasks[worker_id];
+    pub unsafe fn push_task(&self, worker_id: usize, task: ExecutorTask) {
+        self.tasks_size.fetch_add(1, Ordering::SeqCst);
```
Comment (zhang2014) [inline query/src/pipelines/new/executor/executor_tasks.rs:230]: Thread unsafe.

File: query/src/pipelines/new/executor/executor_tasks.rs (hunk 1 above)
Comment (xudong963) [reply query/src/pipelines/new/executor/executor_tasks.rs:58]: yes

**Comment by Xuanwo** [comment]
Is this PR still going on? How about closing it until we are ready to start again?

**Pull request closed without merging**